odnoklassniki-android-sdk/src/main/java/ru/ok/android/sdk/Odnoklassniki.java

```diff
@@ -174,30 +177,55 @@ void onTokenResponseReceived(final Bundle result) {
     }
 
     protected final void notifyFailed(final String error) {
-        notifyFailed(mOkListener, error);
-    }
+        if (mOkListener != null) {
+            notifyFailedInUiThread(mOkListener, error);
+        } else {
+            synchronized (deferredDataList) {
+                deferredDataList.add(new DeferredData(error));
 
-    protected final void notifyFailed(final OkListener listener, final String error) {
-        if (listener != null) {
-            OkThreadUtil.executeOnMain(new Runnable() {
-                public void run() {
-                    listener.onError(error);
+                if (mOkListener != null) {
+                    sendDeferredData(mOkListener);
                 }
-            });
+            }
         }
     }
 
-    protected final void notifySuccess(final JSONObject json) {
-        notifySuccess(mOkListener, json);
+    protected final void notifyFailedInUiThread(final OkListener listener, final String error) {
+        OkThreadUtil.executeOnMain(new Runnable() {
+            public void run() {
+                listener.onError(error);
+            }
+        });
     }
 
-    protected final void notifySuccess(final OkListener listener, final JSONObject json) {
-        if (listener != null) {
-            OkThreadUtil.executeOnMain(new Runnable() {
-                public void run() {
-                    listener.onSuccess(json);
+    protected final void notifySuccess(final JSONObject json) {
+        if (mOkListener != null) {
+            notifySuccessInUiThread(mOkListener, json);
+        } else {
+            synchronized (deferredDataList) {
```

Comment:
using ConcurrentLinkedQueue or analogues will speedup and simplify the corresponding code